electrumsv/network.py

```diff
@@ -838,20 +806,24 @@ def _parse_peers_subscribe(self, result):
         # Take reference so account can be unsubscribed asynchronously without conflict
         subs = self._subs_by_account[account]
         async with TaskGroup() as group:
-            account.request_count += len(triples)
-            account.progress_event.set()
             for keyinstance_id, script_type, script_hash in triples:
                 subs.append(script_hash)
                 # Send request even if already subscribed, as our user expects a response
                 # to trigger other actions and won't get one if we swallow it.
                 self._keyinstance_map[script_hash] = keyinstance_id, script_type
                 await group.spawn(self._subscribe_to_script_hash(script_hash))
 
+            # ensure GUI doesn't keep showing synchronizing...(100/101) when there are only 100 subs
+            account.request_count += len(set(triples) - set(subs))
+            account.progress_event.set()
+
             while await group.next_done():
                 account.response_count += 1
                 account.progress_event.set()
-        # A account shouldn't be subscribing the same key twice
-        assert len(set(subs)) == len(subs)
+
+        if len(set(subs)) != len(subs):
+            seen = set(); duplicates = set(x for x in subs if x in seen or seen.add(x))
+            logger.error("subscribed to the same keys twice for: %s", duplicates)
```

Comment:
Shouldn't this raise? I'd rather we bailed when state gets corrupt than continue onwards.

Reply:
If we subscribe to the same key twice, what's the worst that can happen?

If we raise then it will surely crash in the event that ElectrumX gives bad history (triggering a force disconnect) because this `subscribe_to_triples()` is a long-running coroutine that is difficult to cancel when changing to a new ElectrumX server. So you end up calling `subscribe_to_triples()` again for the new server resulting in some duplicates (but only if `subscribe_to_triples` was still incomplete when the bad history was received.

Up to you though. That is the consequence of raising though vs merely logging the error.

Reply:
I prefer we stop the wallet corruption from getting worse, rather than silently compounding it in logged messages no-one ever reads. When any error on this level happens, we should close the wallet from further writes, display an error and exit. To do anything otherwise creates further problems along the lines of people reporting symptoms of the visible corruption down the road caused by the first silent corruption we hid from them and compounded by all related corruption. Etc.

Reply:
Okay. One moment. I'll make it raise.